### PR TITLE
Make Hash#first(n : Int) return subhash

### PR DESCRIPTION
**Pull request by greyblake**
Implement `Hash#first(n : Int)` method to return subhash.

E.g.
```crystal
{"a" => 1, "b" => 2, "c" => 3}.first(2) 
# => {"a" => 1, "b" => 2}
```

### CODE REVIEW

**Comment by jhass** [comment]
I'd be curious about the usecase for that, I still consider depending on the order of a map that much a code smell.

**Comment by greyblake** [comment]
My use case is the follwing. I have hash where key is a name and value is  a rate of entity.
After sorting the hash (there is not such method for hashes, I had to made it up), I want to get top N items

**Comment by jhass** [comment]
Why is it a hash, compared to say an array of tuples?

**Comment by greyblake** [comment]
Well, it can be of course array of tuples, but then again, I will need to convert it back to Hash, because Hash is much easier to work with. I need to get rate(value) by name(key)

**Comment by greyblake** [comment]
@jhass Actually, does Crystal hash guarantee and order of elements? I've realized, that I am not sure about this. If no, this PR does not make sense and you can close it.

**Comment by jhass** [comment]
It does, it maintains a single linked list through the hash items just like Ruby's does.

**Comment by ysbaddaden** [comment]
I'm not sure there is any **guarantee** for hash elements to be ordered, I would say this is an implementation detail and shouldn't be relied upon.

**Comment by kumpelblase2** [comment]
But given there's a [first](http://crystal-lang.org/api/Hash.html#first-instance-method) instance method, this _is_ already relied on, isn't it?

**Comment by asterite** [comment]
Hash keeps order of insertion, it's not said in the API docs but it should (I didn't have time to document it yet). That said, we can always reconsider this "feature": it takes more memory and logic to maintain the order, but maybe it's not used that much, I don't know.

I'm not sure about a `#first(n)` method. @greyblake Could we see the code that would use this method?

**Comment by greyblake** [comment]
@asterite 

I am working on a natural language detection tool, based on N-grams:

```crystal
  def self.guess_all(sample_text, top = 5)
    sample_trigrams = LanguageProfileBuilder.build(sample_text)
    dists = Array(Tuple(String, Float64)).new(@@profiles.size)
    @@profiles.each do |code, lang_trigrams|
      dist = Comparer.distance(sample_trigrams, lang_trigrams)
      dists << {code, dist}
    end
    sorted = dists.sort_by { |kv| kv[1] }
    sorted.first(top)
  end
```

So, in this example I use `dists` which is array of Tuples, where 1st element is a language code (e.g. `eng`), the second is a distance that represents how close the `sample_text` to every particular language.

`sorted.first(5)` - is more like a pseudo code (currently there is not support for Array#first, see https://github.com/crystal-lang/crystal/pull/2281.

Finally, I want the method to return Hash of first n-top languages with their distances.

P.S. It's also would be good to have a proper `Hash#sort / sort_by` method, but it's gonna be a separate thing.

**Comment by asterite** [comment]
Can't you do `sorted.take(top).to_h`?

**Comment by greyblake** [comment]
Well, of course I can, I still I would prefer to have this method on Hash.

Maybe, if you don't really like the idea, we can close the PR, until somebody else raises similar quation.  I am OK with it.

**Comment by asterite** [comment]
My main concern is that I don't know what would be more intuitive: should `first` return a Hash, or an Array of tuples? The use case is also contrived, it would mean "give me a hash of the first elements that were inserted into this hash". However, in your case you need the first elements because you sorted them first. And it's cheaper to sort an array of tuples than a hash. And in your code you are sorting by value, not sure how would that translate to sorting into a hash.

Basically, all of the above doubts :-)

**Comment by greyblake** [comment]
@asterite I guess you're right, it's a topic to discuss. Personally for me it would be more natural if it returns hash... But if we take a look at Ruby it returns Array of arrays:
```ruby
 {a: 10, b: 20, c: 30}.first(2)
 => [[:a, 10], [:b, 20]] 
```
So for Crystal it would be Array of tuples (still it is better then Array of arrays) :).

So I you think it will be more efficient to return Array of tuples I am OK with :)

**Comment by asterite** [comment]
@greyblake Let's make it return an array of tuples, like in Ruby. Then I'll merge. Thanks!

**Comment by greyblake** [comment]
OK, thanks, I will, as soon as I figure out this problem on my Debian laptop:) https://github.com/crystal-lang/crystal/issues/1964

**Comment by asterite** [comment]
Now that `Hash` is `Enumerable` and `Enumerable#first(n)` exists and it will return an array of tuple (like in Ruby), there's isn't much more to do here. Well, we could override it to return a `Hash`, but you can can to `first(3).to_h` to get the same behaviour.

**Comment by greyblake** [comment]
Cool! Thanks!
